precompiles/gmp/src/lib.rs

```diff
@@ -187,22 +189,71 @@ where
 			.map_err(|_| revert("Amount overflows balance"))?;
 
 		log::debug!(target: "gmp-precompile", "sending XCM via xtokens::transfer...");
-		let call: orml_xtokens::Call<Runtime> = match user_action {
-			VersionedUserAction::V1(action) => orml_xtokens::Call::<Runtime>::transfer {
-				currency_id,
-				amount,
-				dest: Box::new(action.destination),
-				dest_weight_limit: WeightLimit::Unlimited,
-			},
+		let call: Option<orml_xtokens::Call<Runtime>> = match user_action {
+			VersionedUserAction::V1(action) => {
+				log::debug!(target: "gmp-precompile", "Payload: V1");
+				Some(orml_xtokens::Call::<Runtime>::transfer {
+					currency_id,
+					amount,
+					dest: Box::new(action.destination),
+					dest_weight_limit: WeightLimit::Unlimited,
+				})
+			}
+			VersionedUserAction::V2(action) => {
+				log::debug!(target: "gmp-precompile", "Payload: V2");
+				// if the specified fee is more than the amount being transferred, we'll be nice to
+				// the sender and pay them the entire amount.
+				let fee = action.fee.min(amount_transferred);
+
+				if fee > U256::zero() {
+					let output = Self::call(
+						handle,
+						wrapped_address.into(),
+						solidity::encode_with_selector(
+							TRANSFER_SELECTOR,
+							(Address::from(handle.context().caller), fee),
+						),
+					)?;
+					let transferred: bool = solidity::decode_return_value(&output[..])?;
+
+					if !transferred {
+						return Err(RevertReason::custom("failed to transfer() fee").into());
+					}
+				}
+
+				let fee = fee
+					.try_into()
+					.map_err(|_| revert("Fee amount overflows balance"))?;
+
+				log::debug!(target: "gmp-precompile", "deducting fee from transferred amount {:?} - {:?} = {:?}", amount, fee, (amount - fee));
+
+				let remaining = amount.saturating_sub(fee);
+
+				if Into::<u128>::into(remaining) > 0 {
```

Comment:
```suggestion
				if !remaining.is_zero() {
```

The U256 -> u128 conversion is costly, and useless here

Reply:
Good catch, improved in 1ec83c2